OpenFn/kit
# Logger: Sensitive logging

## Short Description

This PR adds options to "sanitise" output in the logger.

Which basically means: if someone tries to log state in a production environment, the logger should automatically remove or obfuscate it.

This PR adds three options: obfuscate, summarize, or remove. By default it will not attempt to sanitize objects (except for state, where we ALWAYS sanitize config), because the CLI is primarily developer tooling.

See the comment below for examples.

## Related issue

Fixes #314 (and I think #211  can be closed out now too)

## Implementation Details

The logger itself takes a `sanitize` argument with different "policies" for what to do when an object or array is logged.

The CLI accepts this as an input and feeds it through to the logger.

If you try and pass an unknown sanitize value, the CLI will abort in very dramatic circumstances (before running any of the job pipeline). That's because if you want to sanitize your output and you make a typo in the argument, it would be irresponsible to run. You're basically telling us you're trying to run in a sensitive production environment where you don't want to log anything.  In practice, of course, this will never happen because those environments are automated. And if you mis-spell "sanitize" I can't help you. But still, one day this might help someone out.

Note that this doesn't yet affect lightning at all - lightning will need to pass an extra flag to the CLI.

## QA Notes

The best way to test this is to create a job which logs some complex values, and see how the logger handles it with different sanitize options.

First install the dev build of openfn, called `openfnx`. From root, do:

`pnpm install && pnpm install:openfx`

(this will build for you and install a global command called `openfnx`)

Then you'll need a test job to run. Here's a simple one:

```
// job.js
fn((s) => {
  console.log({});
  return s;
});
```

Then run this from the CLI. You probably want to do this from a temp directory lest you accidentally check in stuff to the repo@
```
openfnx job.js -ia common --sanitize=summarize
```
Look out for the `[JOB]` log line - that's the bit emitted by your job code, and the bit that should be sanitized.

You can drop the `-i` flag after you've run the job for the first time (it'll run a bit quicker then)

## Checklist before requesting a review
